File: lib/pollyxt_fmi_func_lib/pollyxt_fmi_display_depolcali.py
import matplotlib.pyplot as plt
from matplotlib.colors import ListedColormap
from matplotlib.dates import DateFormatter, DayLocator, HourLocator, MinuteLocator, date2num
from matplotlib.ticker import MultipleLocator, FormatStrFormatter
from matplotlib import use
use('Agg')
import os, sys
import scipy.io as spio
import numpy as np
from datetime import datetime, timedelta

def celltolist(xtickstr):
    tmp = []
    for iElement in range(0, len(xtickstr)):
        if not len(xtickstr[iElement][0]):
            tmp.append('')
        else:
            tmp.append(xtickstr[iElement][0][0])

    return tmp

def datenum_to_datetime(datenum):
    """
    Convert Matlab datenum into Python datetime.
    :param datenum: Date in datenum format
    :return:        Datetime object corresponding to datenum.
    """
    days = datenum % 1
    hours = days % 1 * 24
    minutes = hours % 1 * 60
    seconds = minutes % 1 * 60
    return datetime.fromordinal(int(datenum)) \
           + timedelta(days=int(days)) \
           + timedelta(hours=int(hours)) \
           + timedelta(minutes=int(minutes)) \
           + timedelta(seconds=round(seconds)) \
            - timedelta(days=366)

def rmext(filename):
    file, _ = os.path.splitext(filename)
    return file

def pollyxt_fmi_display_depolcali(tmpFile, saveFolder):
    '''
    Description
    -----------
    Display the housekeeping data from laserlogbook file.

    Parameters
    ----------
    - tmpFile: the .mat file which stores the housekeeping data.

    Return
    ------ 

    Usage
    -----
    pollyxt_fmi_display_depolcali(tmpFile)

    History
    -------
    2019-01-10. First edition by Zhenping

    Copyright
    ---------
    Ground-based Remote Sensing (TROPOS)
    '''

    if not os.path.exists(tmpFile):
        print('{filename} does not exists.'.format(filename=tmpFile))
        return
    
    # read data
    try:
        mat = spio.loadmat(tmpFile, struct_as_record=True)
        sig_t_p = np.transpose(np.concatenate(mat['sig_t_p']))
        sig_t_m = np.transpose(np.concatenate(mat['sig_t_m']))
        sig_x_p = np.transpose(np.concatenate(mat['sig_x_p']))
        sig_x_m = np.transpose(np.concatenate(mat['sig_x_m']))
        wavelength = mat['wavelength'][0][0]
        time = np.concatenate(mat['time'])
        height = np.concatenate(mat['height'])
        caliHIndxRange = np.concatenate(mat['caliHIndxRange'])
        indx_45m = np.concatenate(mat['indx_45m'])
        indx_45p = np.concatenate(mat['indx_45p'])
        dplus = np.transpose(np.concatenate(mat['dplus']))
        dminus = np.transpose(np.concatenate(mat['dminus']))
        segmentLen = np.concatenate(mat['segmentLen'])
        indx = np.concatenate(mat['indx'])
        mean_dplus_tmp = np.concatenate(mat['mean_dplus_tmp'])
        std_dplus_tmp = np.concatenate(mat['std_dplus_tmp'])
        mean_dminus_tmp = np.concatenate(mat['mean_dminus_tmp'])
        std_dminus_tmp = np.concatenate(mat['std_dminus_tmp'])
        TRt = np.concatenate(mat['TR_t'])
        TRx = np.concatenate(mat['TR_x'])
        segIndx = np.concatenate(mat['segIndx'])
        thisCaliTime = np.concatenate(mat['thisCaliTime'])
        pollyVersion = mat['taskInfo']['pollyVersion'][0][0][0]
        location = mat['campaignInfo']['location'][0][0][0]
        version = mat['processInfo']['programVersion'][0][0][0]
    except Exception as e:
        print('%s has been destroyed' % (tmpFile))
        return

    # display
    fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(8, 8), sharey=True, gridspec_kw={'width_ratios': [1, 1]})

    # display signal
    p1, = ax1.semilogx(sig_t_p, height, color='#E75A17', linestyle='-', label=r'$Sig_{+45^\circ}$')
    p2, = ax1.semilogx(sig_t_m, height, color='#1770E7', linestyle='-', label=r'$Sig_{-45^\circ}$')
    p3, = ax1.semilogx(sig_x_p, height, color='#E75A17', linestyle='--', label=r'$Sig_{+45^\circ}$')
    p4, = ax1.semilogx(sig_x_m, height, color='#1770E7', linestyle='--', label=r'$Sig_{-45^\circ}$')
    ax1.set_ylim([height[caliHIndxRange[0] - 1], height[caliHIndxRange[1] - 1]])
    ax1.set_xlim([1, 1e4])
    ax1.set_ylabel('Height (m)')
    ax1.set_xlabel('Signal (a.u.)')
    ax1.yaxis.set_major_locator(MultipleLocator(200))
    ax1.yaxis.set_minor_locator(MultipleLocator(50))
    start = datenum_to_datetime(time[indx_45p[0] - 1])
    end = datenum_to_datetime(time[indx_45m[-1] + 1])
    fig.text(0.5, 0.98, 'Depolarization Calibration for {wave}nm at {start}-{end}'.format(wave=wavelength, start=start.strftime('%H:%M'), end=end.strftime('%H:%M')), horizontalalignment='center', fontweight='bold', fontsize=14)
    l = ax1.legend(handles=[p1, p2, p3, p4], loc='upper right', fontsize=10)

    p1, = ax2.plot(dplus, height[(caliHIndxRange[0] - 1):(caliHIndxRange[1])], color='#E75A17', label=r'$Ratio_{+45^\circ}$')
    p2, = ax2.plot(dminus, height[(caliHIndxRange[0] - 1):(caliHIndxRange[1])], color='#1770E7', label=r'$Ratio_{-45^\circ}$')
    p3, = ax2.plot([0, 1e10], [height[indx + caliHIndxRange[0] - 2], height[indx + caliHIndxRange[1] - 1]], linestyle='--', color='#000000')
    p4, = ax2.plot([0, 1e10], [height[indx + segmentLen + caliHIndxRange[0] - 2], height[indx + segmentLen + caliHIndxRange[1] - 1]], linestyle='--', color='#000000')
    ax2.set_xlim([0, 50])
    ax2.set_xlabel('Ratio')
    l = ax2.legend(handles=[p1, p2], loc='upper right', fontsize=10)
    ax2.text(0.2, 0.7, '$mean_{dplus}=%5.2f, std_{dplus}=%5.3f$\n$mean_{dminus}=%5.2f, std_{dminus}=%5.3f$\n$K=%6.4f, std_K=%6.4f$' % (mean_dplus_tmp[segIndx - 1], std_dplus_tmp[segIndx - 1], mean_dminus_tmp[segIndx - 1], std_dminus_tmp[segIndx - 1], (1 + TRt)/(1 + TRx)*np.sqrt(mean_dplus_tmp[segIndx-1]*mean_dminus_tmp[segIndx - 1]), (1 + TRt)/(1 + TRx) / np.sqrt(mean_dplus_tmp[segIndx - 1] * mean_dminus_tmp[segIndx - 1]) * 0.5 * (mean_dplus_tmp[segIndx - 1] * std_dminus_tmp[segIndx - 1] + mean_dminus_tmp[segIndx - 1] * std_dplus_tmp[segIndx - 1])), fontsize=9, transform=ax2.transAxes)

    fig.text(0.9, - 0.1, '{location}\n{instrument}\nVersion {version}'.format(location=location, instrument=pollyVersion, version=version), fontweight='bold', fontsize=10)

    plt.tight_layout()
    caliTime = datenum_to_datetime(thisCaliTime[0])
    plt.savefig(os.path.join(saveFolder, '{start}_DepolCali_{wave}.png'.format(start=caliTime.strftime('%Y%m%d-%H%M'), wave=wavelength)), dpi=150)
    plt.close()
 
def main():
    pollyxt_fmi_display_depolcali('C:\\Users\\zhenping\\Desktop\\Picasso\\tmp\\tmp.mat', 'C:\\Users\\zhenping\\Desktop\\Picasso\\recent_plots\\POLLYXT_FMI\\20180517')

if __name__ == '__main__':
    # main()
    pollyxt_fmi_display_depolcali(sys.argv[1], sys.argv[2])

File: lib/pollyxt_fmi_func_lib/pollyxt_fmi_display_depolcali.m
function [] = pollyxt_fmi_display_depolcali(data, taskInfo, attri)
%pollyxt_fmi_display_depolcali display the depolarization calibration results
%   Example:
%       [] = pollyxt_fmi_display_depolcali(attri)
%   Inputs:
%       attri
%   Outputs:
%       
%   History:
%       2018-12-29. First Edition by Zhenping
%   Contact:
%       zhenping@tropos.de

global processInfo defaults campaignInfo

if strcmpi(processInfo.visualizationMode, 'matlab')
	%% 355 nm
	% interate over the cali periods
	for iCali = 1:length(attri.depCalAttri355.thisCaliTime)  
		wavelength = 355; 
		time = data.mTime;
		alt = data.height;
		sig_t_p = attri.depCalAttri355.sig_t_p{iCali};
		sig_t_m = attri.depCalAttri355.sig_t_m{iCali};
		sig_x_p = attri.depCalAttri355.sig_x_p{iCali};
		sig_x_m = attri.depCalAttri355.sig_x_m{iCali};
		caliHIndxRange = attri.depCalAttri355.caliHIndxRange{iCali};
		indx_45m = attri.depCalAttri355.indx_45m{iCali};
		indx_45p = attri.depCalAttri355.indx_45p{iCali};
		dplus = attri.depCalAttri355.dplus{iCali};
		dminus = attri.depCalAttri355.dminus{iCali};
		segmentLen = attri.depCalAttri355.segmentLen{iCali};
		indx = attri.depCalAttri355.indx{iCali};
		mean_dplus_tmp = attri.depCalAttri355.mean_dplus_tmp{iCali};
		std_dplus_tmp = attri.depCalAttri355.std_dplus_tmp{iCali};
		mean_dminus_tmp = attri.depCalAttri355.mean_dminus_tmp{iCali};
		std_dminus_tmp = attri.depCalAttri355.std_dminus_tmp{iCali};
		TR_t = attri.depCalAttri355.TR_t{iCali};
		TR_x = attri.depCalAttri355.TR_x{iCali};
		segIndx = attri.depCalAttri355.segIndx{iCali};
		thisCaliTime = attri.depCalAttri355.thisCaliTime{iCali};

		fileDepolCali355 = fullfile(processInfo.pic_folder, taskInfo.pollyVersion, datestr(data.mTime(1), 'yyyymmdd'), sprintf('%s_%d_DepolCali_355.png', datestr(thisCaliTime, 'yyyymmdd-HHMM'), wavelength));

		% visualize calibration process
		figure('position', [0, 0, 600, 600], 'Units', 'Pixels', 'visible', 'off');

		subplot(121);
		p1 = semilogx(sig_t_p, alt, '-b', 'LineWidth', 1, 'DisplayName', 'Sig_{el, +45\circ}'); hold on;
		p2 = semilogx(sig_t_m, alt, '--b', 'LineWidth', 1, 'DisplayName', 'Sig_{el, -45\circ}');
		p3 = semilogx(sig_x_p, alt, '-r', 'LineWidth', 1, 'DisplayName', 'Sig_{x, +45\circ}'); 
		p4 = semilogx(sig_x_m, alt, '--r', 'LineWidth', 1, 'DisplayName', 'Sig_{x, -45\circ}');
		ylim(alt(caliHIndxRange));
		ylabel('Height (m)');
		xlabel('Signal (.a.u)');
		text(1.2, 1.03, sprintf('Depolarization Calibration for %dnm at %s - %s', wavelength, datestr(data.mTime(indx_45p(1) - 1), 'yyyymmdd HH:MM'), datestr(time(indx_45m(end) + 1), 'HH:MM')), 'fontweight', 'bold', 'Units', 'Normal', 'HorizontalAlignment', 'center');
		grid();
		set(gca, 'xminortick', 'on', 'YMinorTick', 'on');
		l = legend([p1, p2, p3, p4], 'Location', 'NorthEast');
		set(l, 'FontSize', 6);

		subplot(122);
		p1 = plot(dplus, alt(caliHIndxRange(1):caliHIndxRange(2)), '-b', 'LineWidth', 1, 'DisplayName', 'Ratio_{+45\circ}'); hold on;
		p2 = plot(dminus, alt(caliHIndxRange(1):caliHIndxRange(2)), '-r', 'LineWidth', 1, 'DisplayName', 'Ratio_{-45\circ}'); hold on;
		plot([0, 1e10], alt([indx + caliHIndxRange(1) - 1, indx + caliHIndxRange(1) - 1]), '--k');
		plot([0, 1e10], alt([indx + segmentLen + caliHIndxRange(1) - 1, indx + segmentLen + caliHIndxRange(1) - 1]), '--k');
		ylim(alt(caliHIndxRange));
		xlim([0.1*min([dplus; dminus]), 3*max([dplus; dminus])]);
		xlabel('Ratio');
		set(gca, 'xminortick', 'on', 'YMinorTick', 'on');
		text(0.2, 0.8, sprintf('mean_{dplus}=%6.2f, std_{dplus}=%5.2f\nmean_{dminus}=%6.2f, std_{dminus}=%5.2f\nK=%6.4f, delta_K=%8.6f', ...
				mean_dplus_tmp(segIndx), std_dplus_tmp(segIndx), mean_dminus_tmp(segIndx), ...
				std_dminus_tmp(segIndx), (1 + TR_t) ./ (1 + TR_x) .* sqrt(mean_dplus_tmp(segIndx) .* mean_dminus_tmp(segIndx)), ...
				(1 + TR_t) ./ (1 + TR_x) ./ sqrt(mean_dplus_tmp(segIndx) .* mean_dminus_tmp(segIndx)) .* 0.5 .* (mean_dplus_tmp(segIndx) .* std_dminus_tmp(segIndx) + mean_dminus_tmp(segIndx) .* std_dplus_tmp(segIndx))), ...
				'Units', 'Normalized', 'fontsize', 8);
		grid();
		l = legend([p1, p2], 'Location', 'NorthEast');
		set(l, 'FontSize', 6);

		text(0.67, -0.08, sprintf(['%s' char(10) '%s' char(10) 'Version %s'], campaignInfo.location, taskInfo.pollyVersion, processInfo.programVersion), 'interpreter', 'none', 'units', 'normal', 'fontsize', 7, 'fontweight', 'bold');

		set(findall(gcf, '-property', 'fontname'), 'fontname', 'Times New Roman');

		export_fig(gcf, fileDepolCali355, '-transparent', '-r300');
		close();

	end

	%% 532 nm
	% interate over the cali periods
	for iCali = 1:length(attri.depCalAttri532.thisCaliTime)  
		wavelength = 532; 
		time = data.mTime;
		alt = data.height;
		sig_t_p = attri.depCalAttri532.sig_t_p{iCali};
		sig_t_m = attri.depCalAttri532.sig_t_m{iCali};
		sig_x_p = attri.depCalAttri532.sig_x_p{iCali};
		sig_x_m = attri.depCalAttri532.sig_x_m{iCali};
		caliHIndxRange = attri.depCalAttri532.caliHIndxRange{iCali};
		indx_45m = attri.depCalAttri532.indx_45m{iCali};
		indx_45p = attri.depCalAttri532.indx_45p{iCali};
		dplus = attri.depCalAttri532.dplus{iCali};
		dminus = attri.depCalAttri532.dminus{iCali};
		segmentLen = attri.depCalAttri532.segmentLen{iCali};
		indx = attri.depCalAttri532.indx{iCali};
		mean_dplus_tmp = attri.depCalAttri532.mean_dplus_tmp{iCali};
		std_dplus_tmp = attri.depCalAttri532.std_dplus_tmp{iCali};
		mean_dminus_tmp = attri.depCalAttri532.mean_dminus_tmp{iCali};
		std_dminus_tmp = attri.depCalAttri532.std_dminus_tmp{iCali};
		TR_t = attri.depCalAttri532.TR_t{iCali};
		TR_x = attri.depCalAttri532.TR_x{iCali};
		segIndx = attri.depCalAttri532.segIndx{iCali};
		thisCaliTime = attri.depCalAttri532.thisCaliTime{iCali};

		fileDepolCali532 = fullfile(processInfo.pic_folder, taskInfo.pollyVersion, datestr(data.mTime(1), 'yyyymmdd'), sprintf('%s_%d_DepolCali_532.png', datestr(thisCaliTime, 'yyyymmdd-HHMM'), wavelength));

		% visualize calibration process
		figure('position', [0, 0, 600, 600], 'Units', 'Pixels', 'visible', 'off');

		subplot(121);
		p1 = semilogx(sig_t_p, alt, '-b', 'LineWidth', 1, 'DisplayName', 'Sig_{el, +45\circ}'); hold on;
		p2 = semilogx(sig_t_m, alt, '--b', 'LineWidth', 1, 'DisplayName', 'Sig_{el, -45\circ}');
		p3 = semilogx(sig_x_p, alt, '-r', 'LineWidth', 1, 'DisplayName', 'Sig_{x, +45\circ}'); 
		p4 = semilogx(sig_x_m, alt, '--r', 'LineWidth', 1, 'DisplayName', 'Sig_{x, -45\circ}');
		ylim(alt(caliHIndxRange));
		ylabel('Height (m)');
		xlabel('Signal (.a.u)');
		text(1.2, 1.03, sprintf('Depolarization Calibration for %dnm at %s - %s', wavelength, datestr(data.mTime(indx_45p(1) - 1), 'yyyymmdd HH:MM'), datestr(time(indx_45m(end) + 1), 'HH:MM')), 'fontweight', 'bold', 'Units', 'Normal', 'HorizontalAlignment', 'center');
		grid();
		set(gca, 'xminortick', 'on', 'YMinorTick', 'on');
		l = legend([p1, p2, p3, p4], 'Location', 'NorthEast');

		subplot(122);
		p1 = plot(dplus, alt(caliHIndxRange(1):caliHIndxRange(2)), '-b', 'LineWidth', 1, 'DisplayName', 'Ratio_{+45\circ}'); hold on;
		p2 = plot(dminus, alt(caliHIndxRange(1):caliHIndxRange(2)), '-r', 'LineWidth', 1, 'DisplayName', 'Ratio_{-45\circ}'); hold on;
		plot([0, 1e10], alt([indx + caliHIndxRange(1) - 1, indx + caliHIndxRange(1) - 1]), '--k');
		plot([0, 1e10], alt([indx + segmentLen + caliHIndxRange(1) - 1, indx + segmentLen + caliHIndxRange(1) - 1]), '--k');
		ylim(alt(caliHIndxRange));
		xlim([0.1*min([dplus; dminus]), 3*max([dplus; dminus])]);
		xlabel('Ratio');
		set(gca, 'xminortick', 'on', 'YMinorTick', 'on');
		text(0.2, 0.8, sprintf('mean_{dplus}=%6.2f, std_{dplus}=%5.2f\nmean_{dminus}=%6.2f, std_{dminus}=%5.2f\nK=%6.4f, delta_K=%8.6f', ...
				mean_dplus_tmp(segIndx), std_dplus_tmp(segIndx), mean_dminus_tmp(segIndx), ...
				std_dminus_tmp(segIndx), (1 + TR_t) ./ (1 + TR_x) .* sqrt(mean_dplus_tmp(segIndx) .* mean_dminus_tmp(segIndx)), ...
				(1 + TR_t) ./ (1 + TR_x) ./ sqrt(mean_dplus_tmp(segIndx) .* mean_dminus_tmp(segIndx)) .* 0.5 .* (mean_dplus_tmp(segIndx) .* std_dminus_tmp(segIndx) + mean_dminus_tmp(segIndx) .* std_dplus_tmp(segIndx))), ...
				'Units', 'Normalized', 'fontsize', 8);
		grid();
		legend([p1, p2], 'Location', 'NorthEast');

		text(0.67, -0.08, sprintf(['%s' char(10) '%s' char(10) 'Version %s'], campaignInfo.location, taskInfo.pollyVersion, processInfo.programVersion), 'interpreter', 'none', 'units', 'normal', 'fontsize', 7, 'fontweight', 'bold');

		set(findall(gcf, '-property', 'fontname'), 'fontname', 'Times New Roman');

		export_fig(gcf, fileDepolCali532, '-transparent', '-r300');
		close();

	end
elseif strcmpi(processInfo.visualizationMode, 'python')
        
    fprintf('Display the results with Python.\n');
    pyFolder = fileparts(mfilename('fullpath'));
    tmpFolder = fullfile(parentFolder(mfilename('fullpath'), 3), 'tmp');
    saveFolder = fullfile(processInfo.pic_folder, taskInfo.pollyVersion, datestr(data.mTime(1), 'yyyymmdd'));

    % create tmp folder by force, if it does not exist.
    if ~ exist(tmpFolder, 'dir')
        fprintf('Create the tmp folder to save the temporary results.\n');
        mkdir(tmpFolder);
    end

	%% interate over the cali periods
	% 532 nm
	for iCali = 1:length(attri.depCalAttri532.thisCaliTime)  
		wavelength = 532; 
		time = data.mTime;
		height = data.height;
		figDPI = processInfo.figDPI;
		sig_t_p = attri.depCalAttri532.sig_t_p{iCali};
		sig_t_m = attri.depCalAttri532.sig_t_m{iCali};
		sig_x_p = attri.depCalAttri532.sig_x_p{iCali};
		sig_x_m = attri.depCalAttri532.sig_x_m{iCali};
		caliHIndxRange = attri.depCalAttri532.caliHIndxRange{iCali};
		indx_45m = attri.depCalAttri532.indx_45m{iCali};
		indx_45p = attri.depCalAttri532.indx_45p{iCali};
		dplus = attri.depCalAttri532.dplus{iCali};
		dminus = attri.depCalAttri532.dminus{iCali};
		segmentLen = attri.depCalAttri532.segmentLen{iCali};
		indx = attri.depCalAttri532.indx{iCali};
		mean_dplus_tmp = attri.depCalAttri532.mean_dplus_tmp{iCali};
		std_dplus_tmp = attri.depCalAttri532.std_dplus_tmp{iCali};
		mean_dminus_tmp = attri.depCalAttri532.mean_dminus_tmp{iCali};
		std_dminus_tmp = attri.depCalAttri532.std_dminus_tmp{iCali};
		TR_t = attri.depCalAttri532.TR_t{iCali};
		TR_x = attri.depCalAttri532.TR_x{iCali};
		segIndx = attri.depCalAttri532.segIndx{iCali};
		thisCaliTime = attri.depCalAttri532.thisCaliTime{iCali};

	    %% display rcs 
	    save(fullfile(tmpFolder, 'tmp.mat'), 'figDPI', 'wavelength', 'time', 'height', 'sig_t_p', 'sig_t_m', 'sig_x_p', 'sig_x_m', 'caliHIndxRange', 'indx_45m', 'indx_45p', 'dplus', 'dminus', 'segmentLen', 'indx', 'mean_dplus_tmp', 'std_dplus_tmp', 'mean_dminus_tmp', 'std_dminus_tmp', 'TR_t', 'TR_x', 'segIndx', 'thisCaliTime', 'processInfo', 'campaignInfo', 'taskInfo');
	    tmpFile = fullfile(tmpFolder, 'tmp.mat');
	    flag = system(sprintf('%s %s %s %s', fullfile(processInfo.pyBinDir, 'python'), fullfile(pyFolder, 'pollyxt_fmi_display_depolcali.py'), tmpFile, saveFolder));
	    if flag ~= 0
	        warning('Error in executing %s', 'pollyxt_fmi_display_depolcali.py');
	    end
	    delete(fullfile(tmpFolder, 'tmp.mat'));
	end

	% 355 nm
	for iCali = 1:length(attri.depCalAttri355.thisCaliTime)  
		wavelength = 355; 
		time = data.mTime;
		height = data.height;
		figDPI = processInfo.figDPI;
		sig_t_p = attri.depCalAttri355.sig_t_p{iCali};
		sig_t_m = attri.depCalAttri355.sig_t_m{iCali};
		sig_x_p = attri.depCalAttri355.sig_x_p{iCali};
		sig_x_m = attri.depCalAttri355.sig_x_m{iCali};
		caliHIndxRange = attri.depCalAttri355.caliHIndxRange{iCali};
		indx_45m = attri.depCalAttri355.indx_45m{iCali};
		indx_45p = attri.depCalAttri355.indx_45p{iCali};
		dplus = attri.depCalAttri355.dplus{iCali};
		dminus = attri.depCalAttri355.dminus{iCali};
		segmentLen = attri.depCalAttri355.segmentLen{iCali};
		indx = attri.depCalAttri355.indx{iCali};
		mean_dplus_tmp = attri.depCalAttri355.mean_dplus_tmp{iCali};
		std_dplus_tmp = attri.depCalAttri355.std_dplus_tmp{iCali};
		mean_dminus_tmp = attri.depCalAttri355.mean_dminus_tmp{iCali};
		std_dminus_tmp = attri.depCalAttri355.std_dminus_tmp{iCali};
		TR_t = attri.depCalAttri355.TR_t{iCali};
		TR_x = attri.depCalAttri355.TR_x{iCali};
		segIndx = attri.depCalAttri355.segIndx{iCali};
		thisCaliTime = attri.depCalAttri355.thisCaliTime{iCali};

	    %% display rcs 
	    save(fullfile(tmpFolder, 'tmp.mat'), 'figDPI', 'wavelength', 'time', 'height', 'sig_t_p', 'sig_t_m', 'sig_x_p', 'sig_x_m', 'caliHIndxRange', 'indx_45m', 'indx_45p', 'dplus', 'dminus', 'segmentLen', 'indx', 'mean_dplus_tmp', 'std_dplus_tmp', 'mean_dminus_tmp', 'std_dminus_tmp', 'TR_t', 'TR_x', 'segIndx', 'thisCaliTime', 'processInfo', 'campaignInfo', 'taskInfo');
	    tmpFile = fullfile(tmpFolder, 'tmp.mat');
	    flag = system(sprintf('%s %s %s %s', fullfile(processInfo.pyBinDir, 'python'), fullfile(pyFolder, 'pollyxt_fmi_display_depolcali.py'), tmpFile, saveFolder));
	    if flag ~= 0
	        warning('Error in executing %s', 'pollyxt_fmi_display_depolcali.py');
	    end
	    delete(fullfile(tmpFolder, 'tmp.mat'));
	end
else
    error('Unknow visualization mode. Please check the settings in pollynet_processing_chain_config.json');
end

end